src/ProjectSerializer.cpp

```diff
@@ -89,115 +94,261 @@ XO("This recovery file was saved by Audacity 2.3.0 or before.\n"
-   Number ReadBigEndian( wxMemoryInputStream &in )
+   std::string_view CacheString(std::string string)
    {
-      Number result;
-      in.Read( &result, sizeof(result) );
-      auto begin = static_cast<unsigned char*>( static_cast<void*>( &result ) );
-      std::reverse( begin, begin + sizeof( result ) );
-      return result;
+      mStringsCache.emplace_back(std::move(string));
+      return mStringsCache.back();
    }
 
-   // Choose between implementations!
-   static const auto WriteUShort =   IsLittleEndian()
-      ? &WriteLittleEndian<UShort>   : &WriteBigEndian<UShort>;
-   static const auto WriteInt =      IsLittleEndian()
-      ? &WriteLittleEndian<Int>      : &WriteBigEndian<Int>;
-   static const auto WriteLong =     IsLittleEndian()
-      ? &WriteLittleEndian<Long>     : &WriteBigEndian<Long>;
-   static const auto WriteULong =    IsLittleEndian()
-      ? &WriteLittleEndian<ULong>    : &WriteBigEndian<ULong>;
-   static const auto WriteLongLong = IsLittleEndian()
-      ? &WriteLittleEndian<LongLong> : &WriteBigEndian<LongLong>;
-
-   static const auto ReadUShort =   IsLittleEndian()
-      ? &ReadLittleEndian<UShort>   : &ReadBigEndian<UShort>;
-   static const auto ReadInt =      IsLittleEndian()
-      ? &ReadLittleEndian<Int>      : &ReadBigEndian<Int>;
-   static const auto ReadLong =     IsLittleEndian()
-      ? &ReadLittleEndian<Long>     : &ReadBigEndian<Long>;
-   static const auto ReadULong =    IsLittleEndian()
-      ? &ReadLittleEndian<ULong>    : &ReadBigEndian<ULong>;
-   static const auto ReadLongLong = IsLittleEndian()
-      ? &ReadLittleEndian<LongLong> : &ReadBigEndian<LongLong>;
-
-   // Functions to read and write certain lengths -- maybe we will change
-   // our choices for widths or signedness?
-
-   using Length = Int;  // Instead, as wide as size_t?
-   static const auto WriteLength = WriteInt;
-   static const auto ReadLength = ReadInt;
-
-   using Digits = Int;  // Instead, just an unsigned char?
-   static const auto WriteDigits = WriteInt;
-   static const auto ReadDigits = ReadInt;
+   XMLTagHandler* mBaseHandler;
+
+   std::vector<XMLTagHandler*> mHandlers;
+
+   std::string_view mCurrentTagName;
+
+   std::deque<std::string> mStringsCache;
+   AttributesList mAttributes;
+
+   bool mInTag { false };
+};
+
+// template<typename BaseCharType>
+// std::string FastStringConvertFromAscii(const BaseCharType* begin, const BaseCharType* end)
+// {
+// 
+// }
+
+template<typename BaseCharType>
+std::string FastStringConvert(const void* bytes, int bytesCount)
+{
+   constexpr int charSize = sizeof(BaseCharType);
+
+   assert(bytesCount % charSize == 0);
+
+   const auto begin = static_cast<const BaseCharType*>(bytes);
+   const auto end = begin + bytesCount / charSize;
+
+   const bool isAscii = std::all_of(
+      begin, end,
+      [](BaseCharType c)
+      { return static_cast<std::make_unsigned_t<BaseCharType>>(c) < 0x7f; });
+
+   if (isAscii)
+      return std::string(begin, end);
```

Comment:
This looks correct, but I don't understand what economy it gives you.  Still there is one std::string construction.  Are you sure there are in fact even more allocations in the non-ascii path?

Reply:
Around 15-20 percent speedup. C++ standard library fails horribly when it comes to working with strings and locales (nothing new, really. It was this way since the STL was created).  Another approach would be to use alternative converters. It will be a bit slower in general, but much faster if there are lots of non ASCII labels in the project.